notifier/notify_windows.go

```diff
@@ -2,351 +2,342 @@ package main
-			log.Infof("failed to parse key %s from SnoreToast-Response into struct: receivedkey is unknown to Portmaster (%s)", elemSplit[0], elem)
+		// Show notification and delete c notification object
+		rc, _, err := lib.showNotification.Call(notificationTemplate)
+		if int64(rc) == -1 {
+			log.Errorf("notify: failed to show notification: %q", err)
 		}
-	}
+		n.systemID = NotificationID(rc)
 
-	if ret.action == "" {
-		log.Errorf("missing attribute action in SnoreToast Response: %s", in)
+		// Link system id to the notification object
+		notifsByIDLock.Lock()
+		notifsByID[NotificationID(rc)] = n
+		notifsByIDLock.Unlock()
 	}
+}
 
-	if ret.id == "" {
-		log.Errorf("missing attribute id in SnoreToast Response: %s", in)
+// Cancel cancels the notification.
+func (n *Notification) Cancel() {
+	if lib == nil {
+		log.Errorf("notify: notification library not properly loaded")
+		return
 	}
 
-	return ret
-}
+	if !libInitialized {
+		log.Error("notify: not initialized")
+	}
 
-func (n *Notification) buildSnoreToastButtonArgument() string {
-	n.Lock()
-	defer n.Unlock()
+	// Lock notification
+	n.RLock()
+	defer n.RUnlock()
 
-	temp := make([]string, 0, len(n.AvailableActions))
-	for _, action := range n.AvailableActions {
-		if !action.IsSupported() {
-			continue
-		}
+	// Lock library
+	lib.Lock()
+	defer lib.Unlock()
 
-		if err := verifySnoreToastArgumentSyntax(action.Text); err != nil {
-			log.Errorf("failed to build SnoreToast Button-Argument: failed to validate Text for %+v: %s", action, err)
-			continue
-		}
+	// the hide function deletes the notification
+	_, _, _ = lib.hideNotification.Call(uintptr(n.systemID))
 
-		if action.Text != "" {
-			temp = append(temp, action.Text)
-		}
-	}
-	return strings.Join(temp, ";")
+	notifsByIDLock.Lock()
+	delete(notifsByID, n.systemID)
+	notifsByIDLock.Unlock()
 }
 
-func runSnoreToastCmd(cmd *exec.Cmd) {
-	exit, err := execCmd(cmd)
+func actionListener() {
+	_ = initialize()
+	// making sure that everything is loaded
+	libInitialized = isInitialized()
+	// Block until notified
+	<-mainCtx.Done()
+}
 
-	switch exit {
-	case 0, 1, 2, 3, 4, 5: // do nothing
-	default:
-		log.Errorf("executing %+v failed: %s", cmd.Args, err)
+// portmasterNotificationCallback is a callback from the c library
+func portmasterNotificationActivatedCallback(id NotificationID, actionIndex int32) uintptr {
+	// The user clicked on the notification (not a button), open the portmaster
+	if actionIndex == -1 {
+		launchApp()
+		return 0
 	}
-}
 
-// Generel helper-Functions:
+	// The user click one of the buttons
+	// Lock for notification map
+	notifsByIDLock.Lock()
+	defer notifsByIDLock.Unlock()
 
-func (list *cmdArgs) addKeyVal(key, val string, required bool) error {
-	if val == "" {
-		if required {
-			return fmt.Errorf("required value for %s is empty", key)
-		}
-		return nil
-	}
+	// Get notified object
+	notification := notifsByID[id]
 
-	*list = append(*list, key, val)
+	notification.Lock()
+	defer notification.Unlock()
```

Comment:
And this why I hate double locking: You lock in the other "direction" this time. Just a question of time until this deadlocks.

In this case, you can just wrap the notifsByIDLock inside a `func(){}()` in order to still be able to use defer unlocking, but don't have to double lock.

Also, check if the notification exists before using it.